packages/api/src/pins.js

```diff
@@ -0,0 +1,86 @@
+import { JSONResponse } from './utils/json-response.js'
+import { normalizeCid } from './utils/normalize-cid.js'
+
+/**
+ * @param {import('./user').AuthenticatedRequest} request
+ * @param {import('./env').Env} env
+ * @param {import('./index').Ctx} ctx
+ */
+export async function pinPost (request, env, ctx) {
+  const { cid, name, origins, meta } = await request.json()
+
+  if (!cid) {
+    return new JSONResponse(
+      { error: { reason: `Invalid request id: ${cid}` }},
+      { status: 400 }
+    )
+  }
+
+  try {
+    const normalizedCid = normalizeCid(cid)
+  } catch (err) {
+    return new JSONResponse(err)
+  }
+
+  // Validate name
+  if (name && typeof name !== 'string') {
+    return new JSONResponse(
+      { error: { reason: 'INVALID_PIN_DATA', details: 'Invalid name' } },
+      { status: 400 }
+    )
+  }
+
+  // Validate origins
+  if (origins && !Array.isArray(origins)) {
+    return new JSONResponse(
+      { error: { reason: 'INVALID_PIN_DATA', details: 'Invalid origins' } },
+      { status: 400 }
+    )
+  }
+
+  // Validate meta
+  if (meta) {
+    const res = new JSONResponse(
+      { error: { reason: 'INVALID_PIN_DATA', details: 'Invalid metadata' } },
```

Comment:
Could we provide a bit more detail to the user about what is wrong with the metadata based on the scenario we're in?